js/components/meter.js

```diff
@@ -0,0 +1,210 @@
+      this.props.onMeterChange(val);
+    }
+  }
+
+  describeArc(cx, cy, radius, angleStart, angleEnd = 0){
+    // arc draws backwards, so start is the offset point, end is at 0%
+    var start = this.getPoint(cx, cy, radius, angleStart);
+    var end = this.getPoint(cx, cy, radius, angleEnd);
+    var d = [
+          "M", start.x, start.y,
+          "A", radius, radius, 0, 0, 0, end.x, end.y
+      ].join(" ");
+      return d;
+  }
+
+  getPoint(cx, cy, r, angle){
+    var rad = (angle-180) * Math.PI / 180;
+    var dx = cx + (r * Math.cos(rad));
+    var dy = cy + (r * Math.sin(rad));
+    var point = {x: dx, y: dy};
+    return point;
+  }
+
+  drawMeterLine(cx, cy, length, angle, pointerWidth){
+    var end = this.getPoint(cx, cy, length, angle);
+    var d = [
+      "M", cx, cy,
+      "L", end.x, end.y,
+      "Z"
+    ].join(" ");
+    return d;
+  }
+
+  generateSegments() {
+    const {cx, cy, r, showSlider, segments } = this.props;
+    let arcSegments = [];
+    for (var i = 0; i < segments.length; i++) {
+      let segmentId = "arc-s" + i;
+      let angleStart = segments[i].end;
+      let angleEnd = segments[i].start;
+      arcSegments.push(<path id={segmentId} key={segmentId} fill="none" stroke={segments[i].color} strokeWidth="34" d={this.describeArc(cx, cy, 80, angleStart, angleEnd)} />);
+    }
+    return arcSegments;
+  }
+
+  startDragging(event) {
+    if (this.props.draggable) {
+      let targetRect = event.currentTarget.getBoundingClientRect();
+      let centerX = (targetRect.width / 2) + targetRect.left;
+      this.setState({ centerX: centerX, minX: centerX - this.props.r, maxX: centerX + this.props.r });
+      let clampedX = this.checkPosition(event.clientX);
+      if (clampedX) {
+        this.updateMeterPosition(clampedX);
+      }
+    }
+    event.preventDefault();
+  }
+  checkPosition(pos) {
+    let {centerX, minX, maxX} = this.state;
+    let r = this.props.r;
+    let padding = 10;
+
+    if (centerX && pos > (minX - padding) && pos < (maxX + padding)) {
+      // clamp position for meter controls while allowing a slight padded tolerance on dragging
+      let clampedPos = pos < minX ? minX : pos > maxX ? maxX : pos;
+      return clampedPos;
+    } else if (centerX) {
+      // was dragging, but now out of bounds
+      this.endDrag();
+    }
+    // not dragging
+    return false;
+  }
+
+  updateMeterPosition(pos) {
+    let minX = this.state.minX;
+    let r = this.props.r;
+    let val = (pos - minX) / (r * 2);
+    this.setMeterValue(val);
+  }
+
+  onDrag(event) {
+    if (this.props.draggable && this.state.centerX) {
+      let clampedX = this.checkPosition(event.clientX);
+      if (clampedX) {
+        this.updateMeterPosition(clampedX);
+      }
+    }
+  }
+  endDrag() {
+    this.setState({centerX: undefined, minX: undefined, maxX: undefined});
+  }
+  finishDragging(event) {
+    this.endDrag();
+    event.preventDefault();
+  }
+
+  render() {
+    const {meterValue} = this.state;
```

Comment:
It seems like instead of calling `scaleCurrentValue` in the constructor and in an `componentWillReceiveProps`, and even having a `meterValue` state property, this could be simplified by just putting `let meterValue = scaleCurrentValue(currentValue);` right here.